docs/contracts/v4/guides/14-accessing-msg.sender-using-hook.mdx

```diff
@@ -0,0 +1,144 @@
+                afterDonate: false,
+                beforeSwapReturnDelta: false,
+                afterSwapReturnDelta: false,
+                afterAddLiquidityReturnDelta: false,
+                afterRemoveLiquidityReturnDelta: false
+            });
+    }
+```
+
+The above code does the following:
+
+* import the relevant dependencies
+* initialize the constructor by passing in the instance of PoolManager
+* override getHookPermissions from BaseHook.sol to return a struct of permissions to signal which hook functions are to be implemented. It will also be used at deployment to validate the address correctly represents the expected permissions.
+
+## Define an Interface for Trusted Routers
+Each trusted router should implement a standard function that exposes the original ```msg.sender```
+
+```
+interface IMsgSender {
+    function msgSender() external view returns (address);
+}
+```
+This function allows hooks to query the router for the actual sender.
+
+## Store a List of Trusted Routers
+The hook should keep track of which router contracts can be trusted to return a valid ```msgSender()```
+
+This can be done with the help of add and remove functions implemented in the hook.
+
+```
+function addRouter(address _router) external {
+        verifiedRouters[_router] = true;
+        emit RouterAdded(_router);
+    }
+```
+
+This function allows hook to add the router to the list of trusted routers.
+
+```
+function removeRouter(address _router) external {
+        verifiedRouters[_router] = false;
+        emit RouterRemoved(_router);
+    }
+```
+
+This function allows hook to remove the router from the list of trusted routers if it's no longer needed.
+
+**<Ins>NOTE:</Ins>** Make sure you include an address mapping in your hook for the routers before adding the functions.
+```
+mapping(address => bool) private verifiedRouters;
+```
+
+
+## Implementing _beforeSwap
+
+Now that we have implemented a basic hook and have added necessary functions, let us implement the beforeSwap function:
+
+```
+function _beforeSwap(
+        address sender,
+        PoolKey calldata,
+        IPoolManager.SwapParams calldata,
+        bytes calldata
+    ) internal override returns (bytes4, BeforeSwapDelta, uint24) {
+
+        if (verifiedRouters[sender]) {
+            console.log("Swap initiated by account:", IMsgSender(sender).msgSender());
+        } else {
+            console.log("Swap initiated by:", sender);
+        }
+
+        return (BaseHook.beforeSwap.selector, BeforeSwapDelta.wrap(0), 0);
+    }
+```
+
+Notice how we are implementing _beforeSwap rather than beforeSwap like in the case of [building base hook example](https://docs.uniswap.org/contracts/v4/guides/hooks/your-first-hook)?
+
+It is because ```beforeSwap``` is marked as virtual in the IHooks interface and since we have written our custom logic, using _beforeSwap will help us to not mess with Uniswap's required validation checks in ```beforeSwap```
+
```

Comment:
at the very end, lets include a full copy+paste of a working hook with its imports, functions, IMsgSender interface, etc